Support for external machine controller (driver)

prashanth26 opened the pull request:
**What this PR does / why we need it**:
With this PR, new providers can be added out of tree (OOT) without sending in PRs to MCM repository. This OOT follows controller patterns where the provider is expected to implement their own machine controller. However, they can make use of the template machine controller provided as a library.

A sample machine controller for AWS can be found here - https://github.com/prashanth26/machine-controller-aws/commits/spliting-controllers2

---------

So the basic idea is that - The existing path of MCM code flow remains as it is for now. So existing clusters running the MCM would work as is today --> AWS machineClaass --> machine ---> machineSet ---> machineDeployment all managed by the same code path at MCM repo.

Anyone trying to build new providers will make use of [some new libraries](https://github.com/prashanth26/machine-controller-manager/tree/spliting-controllers2/pkg/util/provider) which is already used by the [sampleProvider repo](https://github.com/gardener/machine-controller-manager-provider-sampleprovider) and just implement the 5 basic methods following the documentation guide.

So basically a new provider will have to follow this guide - https://github.com/prashanth26/machine-controller-manager/blob/spliting-controllers2/docs/development/cp_support_new.md. In turn they will build his own MachineController and the MachineSet and MachineDeployment controller will be used from the existing MCM code.

And the new providers will make use of the generic MachineClass instead of ProviderSpecficiMachineClass.
MachineClass ---> machine --> Managed by their MC code
MachineSet ---> MachineDeployment ---> Managed by the core current MCM code (edited) 

In the future we will migrate all provider code in MCM into OOT providers and get rid of in-tree code.

**Which issue(s) this PR fixes**:
Fixes #178 

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator|developer
-->
```improvement operator
Support for OOT machine controller 
```

amshuman-kr commented on pkg/util/provider/machinecontroller/machineclass.go:
```diff
@@ -0,0 +1,225 @@
+	return c.reconcileClustermachineClass(class)
+}
+
+func (c *controller) reconcileClustermachineClass(class *v1alpha1.MachineClass) error {
```
```suggestion
func (c *controller) reconcileClusterMachineClass(class *v1alpha1.MachineClass) error {
```

prashanth26 commented:
> Also, why not move the default implementation in the AWS controller into somewhere under pkg/util/provider except for the construction of the driver.Driver implementation struct?

I was wondering wouldn't something like the ability of the provider to configure their own flags? -  the https://github.com/prashanth26/machine-controller-aws/blob/spliting-controllers2/cmd/machine-controller/app/options/options.go#L82-L113. Wouldn't it be better for providers to be able to control this feature?

prashanth26 commented on cmd/machine-controller-manager/app/options/options.go:
```diff
@@ -51,17 +51,18 @@ func NewMCMServer() *MCMServer {
+			KubeAPIBurst:              30,
+			LeaderElection:            leaderelectionconfig.DefaultLeaderElectionConfiguration(),
+			ControllerStartInterval:   metav1.Duration{Duration: 0 * time.Second},
+			ExternalMachineController: false,
```
Yes, we could do this. However, this would partly depend on how our migration logic would look like. For example if we continue working with machine objects the way they are today and ignore the `machineClassType` field to fetch the machine classes, we might need some logic like this. 

But yes, I could revert this change for now and consider reconciling based on `machineClassType` ref in machine objects and reintroduce this change if necessary for migration. Reverting this commit.

amshuman-kr replied:
Actually, this has less to do with migration logic and more to do with backward compatibility. For backward compatibility, you anyway need to check the machineclass kind. Just extend it to filter out the generic machineclass.

prashanth26 replied:
Yes Amshu, this check kind of already exists today - https://github.com/prashanth26/machine-controller-manager/blob/spliting-controllers2/pkg/controller/machine_util.go#L259-L260. Just reworded the log message a little. 

So the current MCM will not process the genericMachineClass ref objects with the current code as well. I forgot to mention this earlier.

amshuman-kr replied:
A minor improvement, currently the filtering is happening during reconciliation. It might be better to do the filtering before enqueing, no?

prashanth26 replied:
Yes, I think that should be better. Shall make that change.

prashanth26 replied:
I have made the changes here - https://github.com/gardener/machine-controller-manager/pull/460/commits/61da081cbade894e5cd3144bc40b7fc97adf54cc.

prashanth26 commented on cmd/machine-controller-manager/app/options/options.go:
```diff
@@ -113,6 +114,7 @@ func (s *MCMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.SafetyOptions.MachineSafetyAPIServerStatusCheckPeriod.Duration, "machine-safety-apiserver-statuscheck-period", s.SafetyOptions.MachineSafetyAPIServerStatusCheckPeriod.Duration, "Time period (in duration) used to poll for APIServer's health by safety controller")
 	fs.StringVar(&s.NodeConditions, "node-conditions", s.NodeConditions, "List of comma-separated/case-sensitive node-conditions which when set to True will change machine to a failed state after MachineHealthTimeout duration. It may further be replaced with a new machine if the machine is backed by a machine-set object.")
 	fs.StringVar(&s.BootstrapTokenAuthExtraGroups, "bootstrap-token-auth-extra-groups", s.BootstrapTokenAuthExtraGroups, "Comma-separated list of groups to set bootstrap token's \"auth-extra-groups\" field to")
+	fs.BoolVar(&s.ExternalMachineController, "external-machine-controller", s.ExternalMachineController, "Set this flag to true to run the machine controller as a seperate controller")
```
Done.

amshuman-kr commented:
> I was wondering wouldn't something like the ability of the provider to configure their own flags? - the https://github.com/prashanth26/machine-controller-aws/blob/spliting-controllers2/cmd/machine-controller/app/options/options.go#L82-L113. Wouldn't it be better for providers to be able to control this feature?

They could always do it like you have done right now, if they really what to support custom flags. WDYT?

prashanth26 commented on pkg/controller/deployment.go:
```diff
@@ -466,14 +466,6 @@ func (dc *controller) reconcileClusterMachineDeployment(key string) error {
 		return nil
 	}
 
-	if deployment.DeletionTimestamp == nil {
```
I was thinking of keeping management of machineSets & machineDeployments controllers independent of underlying machineClass/secrets. There are several reasons why this would be helpful.

1. Theses controller won't bother about the `kind` of machine classes and their validation and their secrets fields. This helps us as these validations for machine classes may not be available at this controller and are part of machine controller. 

2. This makes sense to me as IMO the higher level controllers don't need to bother about these reference objects (machineClasses & secrets). That was we leave validation of such objects to lower level controllers. Also the machine controller may also be implemented with having the providerSpec inline without the use of reference objects if required (the way community does/used to do). 

3. We have observed operations (deletions) being blocked earlier due to such validations and this would prevent such scenarios.  

Hence have removed this here. WDYT?

prashanth26 commented on pkg/controller/machine.go:
```diff
@@ -111,10 +111,15 @@ func (c *controller) reconcileClusterMachineKey(key string) error {
 
 func (c *controller) reconcileClusterMachine(machine *v1alpha1.Machine) error {
-	klog.V(4).Info("Start Reconciling machine: ", machine.Name)
+	if c.externalMachineController {
```
Done. Removing this check.

prashanth26 commented on pkg/controller/machineset.go:
```diff
@@ -463,12 +463,6 @@ func (c *controller) reconcileClusterMachineSet(key string) error {
 	}
 
 	if machineSet.DeletionTimestamp == nil {
-		// Validate MachineClass
```
Same as above.

prashanth26 commented on pkg/util/provider/driver/driver.go:
```diff
@@ -0,0 +1,145 @@
+
+	//  Secret backing the machineClass object
+	Secret *corev1.Secret
+}
```
We have added this `lastKnownState` in the machine object's status. We pass the entire machine object during create/delete request where this state can be retrived from. And for filling this field, we return the lastKnownState aas a response field which is updated on the machine status object - https://github.com/prashanth26/machine-controller-manager/blob/spliting-controllers2/pkg/apis/machine/types.go#L161-L164. 

Is this what you meant?

amshuman-kr replied:
But wouldn't that mean the drivers can modify any part of the `Machine` including status?

I am in principle fine with this. Just pointing it out.

prashanth26 replied:
Yes, I have left it this way to allow drivers to modify the machine objects if required. Especially helps cases where a few providers (xen) would like to add more functionality (updating machine objects for any purpose) which the current interfaces doesn't allow.

prashanth26 commented:
> > I was wondering wouldn't something like the ability of the provider to configure their own flags? - the https://github.com/prashanth26/machine-controller-aws/blob/spliting-controllers2/cmd/machine-controller/app/options/options.go#L82-L113. Wouldn't it be better for providers to be able to control this feature?
> 
> They could always do it like you have done right now, if they really what to support custom flags. WDYT?

I guess that is true. Maybe I shall move the controller initialization also into a library package like you suggested.

amshuman-kr commented:
@prashanth26 Thanks a lot for the changes. I just had another look. The PR looks almost ready!

Only real open point is the external controller flag.

prashanth26 commented:
> @prashanth26 Thanks a lot for the changes. I just had another look. The PR looks almost ready!
Okay, thank you.

> Only real open point is the external controller flag.
Yes, currently I have reverted the changes and kept it to work as it was earlier. It ignores processing machine objects which have reference to any unknown machineClass kind [here]( https://github.com/prashanth26/machine-controller-manager/blob/spliting-controllers2/pkg/controller/machine_util.go#L259-L260). This should work for now, however might need more thought once we plan to migrate objects like `AWSMachineClass` to `MachineClass`

prashanth26 commented on pkg/util/provider/machinecodes/codes/code_string.go:
```diff
@@ -0,0 +1,68 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This file was copied and modified from the kubernetes/kubernetes project
```
Nice catch. Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9

prashanth26 commented on pkg/util/provider/machinecodes/codes/codes.go:
```diff
@@ -0,0 +1,177 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This file was copied and modified from the kubernetes/kubernetes project
```
Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9

prashanth26 commented on pkg/util/provider/machinecodes/status/status.go:
```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This file was copied and modified from the kubernetes/kubernetes project
```
Nice catch. Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9

prashanth26 commented on pkg/util/provider/machinecodes/status/status.go:
```diff
@@ -0,0 +1,116 @@
+// for additional data to be transmitted via the Details field in the status
+// proto.  gRPC service handlers should return an error created by this
+// package, and gRPC clients should expect a corresponding error to be
+// returned from the RPC call.
```
Yes, i have updated. Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9

prashanth26 commented on pkg/util/provider/machinecontroller/machineclass.go:
```diff
@@ -0,0 +1,225 @@
+	"github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+)
+
+// machineClassKind is used to identify the machineClassKind as AWS
```
Good catch. Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9

prashanth26 commented on pkg/util/provider/machinecontroller/machineclass.go:
```diff
@@ -0,0 +1,225 @@
+// machineClassKind is used to identify the machineClassKind as AWS
+const machineClassKind = "machineClass"
+
+func (c *controller) machineDeploymentToMachineClassDelete(obj interface{}) {
```
Not really. I kept for a future use case if required. However, it can be removed here - Nice catch. Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9.

prashanth26 commented on pkg/util/provider/machinecontroller/machineclass.go:
```diff
@@ -0,0 +1,225 @@
+	}
+}
+
+func (c *controller) machineSetToMachineClassDelete(obj interface{}) {
```
Not really. I kept for a future use case if required. However, it can be removed here - Nice catch. Done - https://github.com/gardener/machine-controller-manager/pull/460/commits/a8db3bff253da9d0be5a0acefd4f89826f6a53d9.

prashanth26 commented:
> Thanks a lot, Prashanth. PR looks really good, overall.
Thank you for that. 

> Curious, as I see major changes in machine-controller flow, can we ensure that the couple of fixes we recently had in existing[in-tree]-machine-controller has been taken care of with the new flow?
The driver-specific fixes are something we will have to adopt in the driver implementations (when we add). However, the fixes in the core logic like a drain, machine reconcile I have tried to realign as and when possible. But I/we might have to go through the recent master changes and adapt it. 

I think one of the drain fixe and don't adapt machine without ID maybe pending on this. Will check these out.

prashanth26 commented on pkg/apis/machine/v1alpha1/types.go:
```diff
@@ -167,6 +167,11 @@ type MachineStatus struct {
+	// LastKnownState can store details of the last known state of the VM by the plugins.
+	// It can be used by future operation calls to determine current infrastucture state
+	// +optional
+	LastKnownState string `json:"lastKnownState,omitempty"`
```
The idea is that this field will hold any cloud provider-specific status/info in this field. The status that is irrelevant to core MCM but the drivers. Helps in cases where machine might be partially created in one call and might be adopted in the next call.

prashanth26 replied:
This probably will be a lot more clearer when I add the documentation (positively today) which explains how new providers can be added.

hardikdr replied:
Sure, thanks for the clarification.

prashanth26 commented:
Hi All, Hi @hardikdr ,

I have updated the documentation to add new CP support using OOT [here in this PR]( https://github.com/prashanth26/machine-controller-manager/blob/spliting-controllers2/docs/development/cp_support_new.md). This is a first draft which should mostly be correct, however, we may improve these documentations over time. 

The [SampleProvider repository](https://github.com/gardener/machine-controller-manager-provider-sampleprovider) has also been updated to reflect the changes. 

Once this PR is merged, I shall align both the AWS and Vsphere providers to vendor this version of MCM controller libraries. 

>Curious, as I see major changes in machine-controller flow, can we ensure that the couple of fixes
we recently had in existing[in-tree]-machine-controller has been taken care of with the new flow?

Will check the latest fixes on MCM and migrate them here by EOD there after which, I think we can merge this PR.

prashanth26 commented:
Hi All,

I am done with all my changes for OOT. I hope i have addressed all your concerns. Please raise any major concerns you may have, minor concerns regarding documentation could probably be addressed with a follow-up PR. We plan to positively merge this PR by EOD and make a release. 

Regards,
Prashanth

prashanth26 commented:
Hi All,

I have just squashed a few commits for better readability. I have also merged the master with the current branch.
